### ISSUE TITLE
Access to counts of tuples written by `writeExecution`.

### ISSUE THREAD

**Issue by rowandavies** [twitter/scalding#1461]
Currently there's no easy and efficient way to access the counts of tuples written by `writeExecution`, which is often useful for testing and detection of variations from expected counts.

At the cascading level, you could use `cascading.flow.StepCounters.Tuples_Written` for the final step in the flow.  But, from scalding there doesn't seem to be an easy way to do the same - you can only get the sum of the counts of all the steps in a flow, unless I've missed something.  This does work when there's only a single step, but not otherwise.

It should work add an additional counter `inc` at the end of such pipes that is essentially just counting the same tuples as `StepCounters.Tuples_Written` for that step, but that seems like duplication- unless this the intended way of doing such things?

**Comment by johnynek** [comment]
Execution.getCounters should do it. Does that not work for you?

On Thursday, December 17, 2015, rowandavies <notifications@github.com>
wrote:

> Currently there's no easy and efficient way to access the counts of tuples
> written by writeExecution, which is often useful for testing and
> detection of variations from expected counts.
>
> At the cascading level, you could use
> cascading.flow.StepCounters.Tuples_Written for the final step in the
> flow. But, from scalding there doesn't seem to be an easy way to do the
> same - you can only get the sum of the counts of all the steps in a flow,
> unless I've missed something. This does work when there's only a single
> step, but not otherwise.
>
> It should work add an additional counter inc at the end of such pipes
> that is essentially just counting the same tuples as
> StepCounters.Tuples_Written for that step, but that seems like
> duplication- unless this the intended way of doing such things?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/issues/1461>.
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin

**Comment by stephanh** [comment]
`Execution.getCounters` gives access to the Cascading counters. However, the cascading counters such as `Tuples_Written, cascading.flow.SliceCounters` accumulate the counters from each step.

For example if you have an execution that gets translated into 3 sequential MR jobs, each MR job writes out 1 tuple and the final write operation writes out 1 tuple, then you get a `Tuples_Written` count of 3 instead of the 1 you might expect.

We can't see any good approaches to use any of the other existing counters to actually find out how many records were written out.

**Comment by johnynek** [comment]
There is also `Execution.getAndReset`, but I'm not actually sure what you want. What would you like to see in the counters?

For instance, the MR-job concept is not present in the scalding API. So, that is an implementation detail. Exposing too much of that really constrains what the code can do and how portable it is (for instance, we want to support Tez and maybe Flink, what should an API that exposes MR job counters do there?).

**Comment by stephanh** [comment]
We currently use `getAndResetCounters`.

@johnynek I think we want the same thing. Execution abstracts away the concept of MR jobs but I would say the current counters break that abstraction barrier.

Here is an example:

```scala
val data = List(
    ("A", "1"),
   ("A", "2"),
   ("B", "1"),
   ("B", "2")
  )

val exec = IterableSource(data)
  .groupBy(_._1).maxBy(_._2).values     // List(("A", 2), ("B", 2))
  .groupBy(_._2).maxBy(_._1).values     // List(("B", 2))
  .writeExecution(TypedPsv("out")

val count = exec.getAndRestCounters.map { (_, counters) => counters(StatKey("Tuples_Written", "cascading.flow.SliceCounters")) }
```

The count is 5 and we would like to get a count with just the number of values that were actually written to `out`, 1 in this case.

There doesn't seem to be a good way to do that apart from incrementing our own counter right at the end.

**Comment by stephanh** [comment]
This is the work around we are currently considering https://github.com/CommBank/maestro/pull/469.

**Comment by rowandavies** [comment]
@johnynek What may make sense is to have some way to fetch the counters like `cascading.flow.StepCounters.Tuples_Written` just for the final step in the flow for a `writeExecution`, or more generally just for steps that do "external" output from the flow, as opposed to steps that do output that is only "internal" to the flow i.e. only to create input for later steps in the flow.  

While scalding shouldn't expose the concept of a step, defining the counter value for a flow as the sum over the steps (as cascading does), including internal steps, seems to actually expose more implementation details, at least for this counter.  (Breaking the abstraction barrier, as Stephan said.)

**Comment by johnynek** [comment]
Certainly, counters are side effects that we chose not to expose in the type-system. We debated that a lot, and the feeling was that they would be too cumbersome to use if we did that, maybe that was wrong.

As a result, yes, your example shows that the counters will break some of the abstractions.

The main optimization I want to preserve is the ability to merge flows together that have not yet run:
https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/Execution.scala#L549

The purpose of which is to let cascading manage scheduling in a hopefully efficient way.

So, by the time that block of code finishes, cascading has already collected and merged the counters.

Is the problem that we only keep the counters here, and not, for instance the flow_step_stats here:
https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/JobStats.scala#L29

This code here:
https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/Execution.scala#L824

just gets the aggregated counters. We need a Monoid on the ExecutionCounters or else we will really have to change the current design. We could imagine some monoid or another on ExecutionCounters that include the flow_step_stats. We could keep an integer that we are incrementing each time we run a flow. Then we could make a `Map[Int, Map[String, Long]]` as the flow_step_stats, which has the `Int` for which flow we have run as the key (but I doubt that id will be deterministic), so it might not be a great idea.

Another idea would be to consider the flow_step_stats as a `Vector[Map[String, Long]]` and use vector-space addition (padding with zero) to add multiple flows together. If you don't merge multiple flows (with zip), then you should see the counters you want.

But your proposed work-around looks fine to me: create your own counter for what you exactly mean. I don't see the problem with that.

Frankly, I really don't like counters and have often thought they were a mistake to put into the Execution API. I think they should only be used for debugging, and never for flow control. But I guess that ship has sailed.

What do you all think? Any other proposals?

**Comment by avibryant** [comment]
This is maybe a bit of a tangent but if we're *going* to allow counters for flow control, we should have a nicer interface to them - seems like we want an interface equivalent to `Aggregator[A, Map[String,Long], C]`, only allowing the default monoid for the intermediate type (since that's the only semantics hadoop gives us).

**Comment by avibryant** [comment]
Alternatively `Injection[A,Map[String,Long]]` which covers any cases I can immediately think of.

**Comment by rowandavies** [comment]
@johnynek I think you're on the right track, and I agree that merging flows is important.  

Counters do seem to be inherently leaky in the context of scalding abstractions, but there seem to be some uses of some counters that are robust to lower-level implementation details and are useful and somewhat natural - in our specific case I think we'd end up with an extra map-reduce step in over 1000 scheduled jobs if we avoided counters altogether.  And, in our case I think the robustness requires a concept like the counters associated with the "writing part" of a `writeExecution` (underneath corresponding to a particular cascading flow step).

Some general interface for combining the `flow_step_stats` for the steps, via a caller provided monoid (or Aggregator as suggested by @avibryant) could be appropriate and could match the current behaviour via a default.  Then, we could adapt the example posted by @stephanh to use a monoid that returns the last value, which will be robust as long as we somehow guarantee that the last element in the sequence of `flow_step_stats`  is for the step at the end of the flow that does the writing.  

Or, more generally something like accumulating a `Map[Int, Map[String, Long]]` should work if there was a variant of `writeExecution` that also yielded the `Int` key corresponding to the part that actually does the writing.  I guess this assumes that there's a robust enough concept of "the part that actually does the writing" that can be simply implemented by directly fetching the counters for the corresponding cascading flow step - I think there is, but I could be wrong.

A simpler approach that would be sufficient for the example and our use case would be to directly add the concept of the counters associated specifically with a use of `writeExecution` on a `TypedPipe`, meaning the counters associated with the writing step, implemented by fetching the counters for the corresponding cascading flow step.  This is clearly less general, but it's unclear whether a more general approach has additional robust uses, or just supports more general forms of abstraction leaks. So maybe taking small steps is reasonable in this area unless/until we find other specific robust uses of sub-flow counters.
